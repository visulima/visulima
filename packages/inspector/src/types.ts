export type Indent = {
    base: string;
    prev: string;
};

export type InternalInspect = (input: unknown, from: unknown, options: Options) => string;

export type InspectType<V> = (input: V, options: Options, inspect: InternalInspect, indent: Indent | undefined) => string;

export type Inspect = (input: unknown, options: Options) => string;

export interface Options {
    breakLength: number;
    customInspect: boolean;
    depth: number;
    indent: number | "\t" | undefined;
    maxArrayLength: number;
    numericSeparator: boolean;
    quoteStyle: "double" | "single";
    showHidden: boolean;
    showProxy: boolean;
    stylize: <S extends string>(
        value: S,
        // eslint-disable-next-line @typescript-eslint/no-redundant-type-constituents
        styleType: string | "bigint" | "boolean" | "date" | "null" | "number" | "regexp" | "special" | "string" | "symbol" | "undefined",
    ) => string;
    truncate: number;
}
