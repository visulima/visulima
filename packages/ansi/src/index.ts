export { default as slice } from "./slice";
