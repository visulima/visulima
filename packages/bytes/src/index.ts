import { Buffer } from "node:buffer";

/**
 * Converts a Node.js Buffer to a Uint8Array.
 * Note: While Buffer instances are Uint8Array instances, this function
 * creates a new Uint8Array view on the Buffer's underlying ArrayBuffer,
 * ensuring correct byteOffset and length.
 * @param buf The Buffer to convert.
 * @returns A Uint8Array view of the Buffer.
 */
export const bufferToUint8Array = (buf: Buffer): Uint8Array => new Uint8Array(buf.buffer, buf.byteOffset, buf.length);

/**
 * Checks if a value is a Uint8Array or, in a Node.js environment, a Buffer.
 * @param x The value to check.
 * @returns True if x is a Uint8Array or Buffer, false otherwise.
 */
export const isUint8Array
  = typeof Buffer === "function"
      ? (x: unknown): x is Uint8Array => x instanceof Uint8Array || Buffer.isBuffer(x)
      : (x: unknown): x is Uint8Array => x instanceof Uint8Array;

/**
 * Converts an ASCII string, array of strings, or template literal to a Uint8Array.
 * Each character is converted to its ASCII byte value (0-255). Characters outside this range will be truncated.
 * @param txt The input string, array of strings, or template strings array.
 * @returns A Uint8Array representing the ASCII encoded string.
 */
export const asciiToUint8Array = (txt: TemplateStringsArray | string | [string]): Uint8Array => {
    if (typeof txt === "string")
        return asciiToUint8Array([txt]);

    const [input] = Array.isArray(txt) ? txt : [String.raw(txt as TemplateStringsArray)]; // Handle TemplateStringsArray
    const inputLength = input.length;
    const result = new Uint8Array(inputLength); // Renamed 'res' to 'result'

    for (let index = 0; index < inputLength; index += 1) { // Changed index++ to index += 1
        // eslint-disable-next-line unicorn/prefer-code-point, no-bitwise
        result[index] = input.charCodeAt(index) & 0xFF; // Ensure ASCII range
    }

    return result;
};

/**
 * Converts a UTF-8 string, array of strings, or template literal to a Uint8Array.
 * Requires Node.js Buffer support.
 * @param txt The input string, array of strings, or template strings array.
 * @returns A Uint8Array representing the UTF-8 encoded string.
 */
export const utf8ToUint8Array = (txt: TemplateStringsArray | [string] | string): Uint8Array => {
    if (typeof txt === "string")
        return utf8ToUint8Array([txt]);

    const [input] = Array.isArray(txt) ? txt : [String.raw(txt as TemplateStringsArray)]; // Handle TemplateStringsArray

    return bufferToUint8Array(Buffer.from(input, "utf8"));
};

/**
 * Attempts to convert various data types to a Uint8Array.
 * Supports Uint8Array (returns as is), ArrayBuffer, Array of numbers.
 * In Node.js, also supports Buffer (returns as is or converts from other types via Buffer.from).
 * @param data The data to convert.
 * @returns A Uint8Array representation of the input data.
 * @throws Error 'UINT8ARRAY_INCOMPATIBLE' if the data cannot be converted and Buffer is not available for fallback.
 */
export const toUint8Array = (data: unknown): Uint8Array => {
    if (typeof Buffer === "function" && Buffer.isBuffer(data)) {
        return bufferToUint8Array(data); // Prioritize Buffer and ensure plain Uint8Array
    }

    if (data instanceof Uint8Array) { // This will now handle non-Buffer Uint8Arrays
        return data;
    }

    if (data instanceof ArrayBuffer)
        return new Uint8Array(data);

    // Ensure it's an array of numbers before creating Uint8Array from it
    if (Array.isArray(data) && data.every((item) => typeof item === "number")) {
        return new Uint8Array(data);
    }

    if (typeof Buffer === "function") {
        if (Buffer.isBuffer(data)) {
            return bufferToUint8Array(data); // Use our converter for consistency
        }

        // Attempt to convert string via Buffer.from.
        // Other Buffer-compatible types like ArrayBuffer or ArrayBufferView (which Buffer is)
        // are typically handled by earlier checks or are Buffers themselves.
        if (typeof data === "string") {
            try {
                return bufferToUint8Array(Buffer.from(data));
            } catch {
                // If Buffer.from fails, fall through to error
            }
        }
    }

    throw new Error("UINT8ARRAY_INCOMPATIBLE: Cannot convert data to Uint8Array");
};

// eslint-disable-next-line import/no-extraneous-dependencies
export * from "@std/bytes";
