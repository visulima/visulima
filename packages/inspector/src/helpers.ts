import type { Inspect, Options } from "./types";

const ansiColors = {
    // 21-29 are resets for 1-9
    black: ["30", "39"],
    blue: ["34", "39"],
    bold: ["1", "22"],
    brightblack: ["30;1", "39"],
    brightblue: ["34;1", "39"],
    brightcyan: ["36;1", "39"],
    brightgreen: ["32;1", "39"],
    // 10-20 are fonts
    brightmagenta: ["35;1", "39"],
    brightred: ["31;1", "39"],
    brightwhite: ["37;1", "39"],
    brightyellow: ["33;1", "39"],
    cyan: ["36", "39"],
    dim: ["2", "22"],
    green: ["32", "39"],
    grey: ["90", "39"],

    hidden: ["8", "28"],
    // 5 & 6 are blinking
    inverse: ["7", "27"],
    italic: ["3", "23"],
    magenta: ["35", "39"],
    red: ["31", "39"],
    strike: ["9", "29"],
    underline: ["4", "24"],
    white: ["37", "39"],

    yellow: ["33", "39"],
} as const;

const styles: Record<string, keyof typeof ansiColors> = {
    bigint: "yellow",
    boolean: "yellow",
    date: "magenta",
    null: "bold",
    number: "yellow",
    regexp: "red",
    special: "cyan",
    string: "green",
    symbol: "green",
    undefined: "grey",
} as const;

export const truncator = "â€¦";

type AnsiValue = (typeof ansiColors)[keyof typeof ansiColors];

function colorise<S extends string>(value: S, styleType: string): string {
    const color: AnsiValue | undefined = ansiColors[styles[styleType]] || ansiColors[styleType as keyof typeof ansiColors] || "";

    if (!color) {
        return String(value) as S;
    }

    return `\u001B[${color[0]}m${String(value) as S}\u001B[${color[1]}m`;
}

export function normaliseOptions(
    {
        breakLength = Number.POSITIVE_INFINITY,
        colors = false,
        customInspect = true,
        depth = 2,
        maxArrayLength = Number.POSITIVE_INFINITY,
        seen = [],
        showHidden = false,
        showProxy = false,
        stylize = String,

        truncate = Number.POSITIVE_INFINITY,
    }: Partial<Options> = {},
    inspect: Inspect,
): Options {
    const options = {
        breakLength: Number(breakLength),
        colors: Boolean(colors),
        customInspect: Boolean(customInspect),
        depth: Number(depth),
        inspect,
        maxArrayLength: Number(maxArrayLength),
        seen,
        showHidden: Boolean(showHidden),
        showProxy: Boolean(showProxy),
        stylize,
        truncate: Number(truncate),
    };
    if (options.colors) {
        options.stylize = colorise;
    }
    return options;
}

function isHighSurrogate(char: string): boolean {
    return char >= "\uD800" && char <= "\uDBFF";
}

export function truncate(string: number | string, length: number, tail: typeof truncator = truncator) {
    string = String(string);
    const tailLength = tail.length;
    const stringLength = string.length;
    if (tailLength > length && stringLength > tailLength) {
        return tail;
    }
    if (stringLength > length && stringLength > tailLength) {
        let end = length - tailLength;
        if (end > 0 && isHighSurrogate(string[end - 1])) {
            end -= 1;
        }
        return `${string.slice(0, end)}${tail}`;
    }
    return string;
}

export function inspectList(list: ArrayLike<unknown>, options: Options, inspectItem?: Inspect, separator = ", "): string {
    inspectItem = inspectItem || options.inspect;
    const size = list.length;
    if (size === 0) return "";
    const originalLength = options.truncate;
    let output = "";
    let peek = "";
    let truncated = "";
    for (let index = 0; index < size; index += 1) {
        const last = index + 1 === list.length;
        const secondToLast = index + 2 === list.length;
        truncated = `${truncator}(${list.length - index})`;
        const value = list[index];

        // If there is more than one remaining we need to account for a separator of `, `
        options.truncate = originalLength - output.length - (last ? 0 : separator.length);
        const string = peek || inspectItem(value, options) + (last ? "" : separator);
        const nextLength = output.length + string.length;
        const truncatedLength = nextLength + truncated.length;

        // If this is the last element, and adding it would
        // take us over length, but adding the truncator wouldn't - then break now
        if (last && nextLength > originalLength && output.length + truncated.length <= originalLength) {
            break;
        }

        // If this isn't the last or second to last element to scan,
        // but the string is already over length then break here
        if (!last && !secondToLast && truncatedLength > originalLength) {
            break;
        }

        // Peek at the next string to determine if we should
        // break early before adding this item to the output
        peek = last ? "" : inspectItem(list[index + 1], options) + (secondToLast ? "" : separator);

        // If we have one element left, but this element and
        // the next takes over length, the break early
        if (!last && secondToLast && truncatedLength > originalLength && nextLength + peek.length > originalLength) {
            break;
        }

        output += string;

        // If the next element takes us to length -
        // but there are more after that, then we should truncate now
        if (!last && !secondToLast && nextLength + peek.length >= originalLength) {
            truncated = `${truncator}(${list.length - index - 1})`;
            break;
        }

        truncated = "";
    }
    return `${output}${truncated}`;
}

function quoteComplexKey(key: string): string {
    if (/^[a-z_]\w*$/i.test(key)) {
        return key;
    }
    return JSON.stringify(key)
        .replaceAll("'", "\\'")
        .replaceAll('\\"', '"')
        .replaceAll(/(^"|"$)/g, "'");
}

export function inspectProperty([key, value]: [unknown, unknown], options: Options): string {
    options.truncate -= 2;
    if (typeof key === "string") {
        key = quoteComplexKey(key);
    } else if (typeof key !== "number") {
        key = `[${options.inspect(key, options)}]`;
    }
    options.truncate -= (key as string).length;
    value = options.inspect(value, options);
    return `${key}: ${value}`;
}
