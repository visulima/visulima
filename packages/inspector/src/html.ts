import { TRUNCATOR } from "./constants";
import type { InternalInspect, Options } from "./types";
import inspectList from "./utils/inspect-list";

const inspectAttribute = ([key, value]: [unknown, unknown], _: unknown, options: Options): string => {
    // eslint-disable-next-line no-param-reassign
    options.truncate -= 3;

    if (!value) {
        return `${options.stylize(String(key), "yellow")}`;
    }

    return `${options.stylize(String(key), "yellow")}=${options.stylize(`"${value as string}"`, "string")}`;
};

export const inspectHTMLCollection = (collection: ArrayLike<Element>, options: Options, inspect: InternalInspect): string =>
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    inspectList(collection, collection, options, inspect, inspectHTMLElement, "\n");

export const inspectHTMLElement = (element: Element, object: unknown, options: Options, inspect: InternalInspect): string => {
    const properties = element.getAttributeNames();
    const name = element.tagName.toLowerCase();
    const head = options.stylize(`<${name}`, "special");
    const headClose = options.stylize(`>`, "special");
    const tail = options.stylize(`</${name}>`, "special");

    // eslint-disable-next-line no-param-reassign
    options.truncate -= name.length * 2 + 5;

    let propertyContents = "";

    if (properties.length > 0) {
        propertyContents += " ";
        propertyContents += inspectList(
            properties.map((key: string) => [key, element.getAttribute(key)]),
            object,
            options,
            inspect,
            inspectAttribute,
            " ",
        );
    }

    // eslint-disable-next-line no-param-reassign
    options.truncate -= propertyContents.length;

    const { truncate } = options;

    let children = inspectHTMLCollection(element.children, options, inspect);

    if (children && children.length > truncate) {
        children = `${TRUNCATOR}(${element.children.length})`;
    }

    return `${head}${propertyContents}${headClose}${children}${tail}`;
};
