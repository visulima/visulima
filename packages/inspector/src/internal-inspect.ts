import { inspectHTMLElement, inspectNodeCollection } from "./html";
import type { Inspect, InspectType, InternalInspect, InternalOptions } from "./types";
import inspectArguments from "./types/arguments";
import inspectArray from "./types/array";
import inspectBigInt from "./types/bigint";
import inspectClass from "./types/class";
import inspectDate from "./types/date";
import inspectError from "./types/error";
import inspectFunction from "./types/function";
import inspectMap from "./types/map";
import inspectNumber from "./types/number";
import inspectObject from "./types/object";
import inspectPromise from "./types/promise";
import inspectRegExp from "./types/regexp";
import inspectSet from "./types/set";
import inspectString from "./types/string";
import inspectSymbol from "./types/symbol";
import inspectTypedArray from "./types/typed-array";
import { getIndent } from "./utils/indent";

// eslint-disable-next-line @typescript-eslint/no-unsafe-function-type
const constructorMap = new WeakMap<Function, Inspect>();
const stringTagMap: Record<string, Inspect> = {};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const baseTypesMap: Record<string, InspectType<any>> = {
    Arguments: inspectArguments,
    Array: inspectArray,

    ArrayBuffer: () => "",
    BigInt: inspectBigInt,

    bigint: inspectBigInt,
    Boolean: (value: boolean, options: InternalOptions) => options.stylize(String(value), "boolean"),
    boolean: (value: boolean, options: InternalOptions) => options.stylize(String(value), "boolean"),

    DataView: () => "",

    Date: inspectDate,
    Error: inspectError,

    Float32Array: inspectTypedArray,
    Float64Array: inspectTypedArray,

    Function: inspectFunction,
    function: inspectFunction,

    Generator: () => "",
    HTMLCollection: inspectNodeCollection,
    Int8Array: inspectTypedArray,
    Int16Array: inspectTypedArray,
    Int32Array: inspectTypedArray,
    Map: inspectMap,

    NodeList: inspectNodeCollection,
    null: (_value: null, options: InternalOptions) => options.stylize("null", "null"),

    Number: inspectNumber,
    number: inspectNumber,
    Promise: inspectPromise,
    RegExp: inspectRegExp,
    Set: inspectSet,
    String: inspectString,
    string: inspectString,
    // A Symbol polyfill will return `Symbol` not `symbol` from typedetect
    Symbol: inspectSymbol,
    symbol: inspectSymbol,
    Uint8Array: inspectTypedArray,

    Uint8ClampedArray: inspectTypedArray,
    Uint16Array: inspectTypedArray,
    Uint32Array: inspectTypedArray,

    undefined: (_value: undefined, options: InternalOptions) => options.stylize("undefined", "undefined"),

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    WeakMap: (_value: WeakMap<any, unknown>, options: InternalOptions) => options.stylize("WeakMap{…}", "special"),
    // WeakSet, WeakMap are totally opaque to us
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    WeakSet: (_value: WeakSet<any>, options: InternalOptions) => options.stylize("WeakSet{…}", "special"),
} as const;

const inspectCustom = (value: object, options: InternalOptions, type: string, inspectFunction_: InternalInspect): string | undefined => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    if (globalThis.window === undefined && typeof (value as any)[Symbol.for("nodejs.util.inspect.custom")] === "function") {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-function-type
        return ((value as any)[Symbol.for("nodejs.util.inspect.custom")] as Function)(options.depth, options, inspectFunction_);
    }

    if ("inspect" in value && typeof value.inspect === "function") {
        return value.inspect(options.depth, options);
    }

    if ("constructor" in value && constructorMap.has(value.constructor)) {
        return constructorMap.get(value.constructor)?.(value, options) ?? "unknown";
    }

    if (stringTagMap[type]) {
        return (stringTagMap[type] as Inspect)(value, options);
    }

    return "";
};

// eslint-disable-next-line sonarjs/cognitive-complexity
export const internalInspect = (value: unknown, options: InternalOptions, depth: number, seen: unknown[]): string => {
    if (seen.includes(value)) {
        return "[Circular]";
    }

    // eslint-disable-next-line @typescript-eslint/no-shadow
    const inspect: InternalInspect = (object: unknown, from: unknown, options: InternalOptions): string => {
        if (from) {
            // eslint-disable-next-line no-param-reassign
            seen = [...seen];
            seen.push(from);
        }

        return internalInspect(object, options, depth + 1, seen);
    };

    const indent = options.indent ? getIndent(options.indent, depth) : undefined;

    const multiline = value && (typeof value === "object" || Array.isArray(value));

    if (options.depth !== undefined && depth >= options.depth && options.depth > 0 && multiline) {
        return Array.isArray(value) ? "[Array]" : "[Object]";
    }

    let type = value === null ? "null" : typeof value;

    if (type === "object") {
        type = Object.prototype.toString.call(value).slice(8, -1);
    }

    if (options.showProxy && options.proxyHandler) {
        return options.proxyHandler(value as typeof Proxy, options, inspect, indent, depth);
    }

    // If it is a base value that we already support, then use inspector
    if (baseTypesMap[type as keyof typeof baseTypesMap] !== undefined) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        return (baseTypesMap[type as keyof typeof baseTypesMap] as InspectType<any>)(value, options, inspect, indent, depth);
    }

    // If `options.customInspect` is set to true then try to use the custom inspector
    if (options.customInspect && value) {
        // eslint-disable-next-line no-param-reassign
        options.depth = options.depth === undefined ? Number.POSITIVE_INFINITY : options.depth - depth;

        const output = inspectCustom(value, options, type, inspect);

        if (output) {
            if (typeof output === "string") {
                return output;
            }

            return inspect(output, value, options);
        }
    }

    const proto = value ? Object.getPrototypeOf(value) : false;

    // If it's a plain Object then use inspector
    if (proto === Object.prototype || proto === null) {
        return inspectObject(value as object, options, inspect, indent, depth);
    }

    // Specifically account for HTMLElements
    if (value && typeof HTMLElement === "function" && value instanceof HTMLElement) {
        return inspectHTMLElement(value, value, options, inspect, depth);
    }

    if ("constructor" in (value as object)) {
        // If it is a class, inspect it like an object but add the constructor name
        if ((value as object).constructor !== Object) {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            return inspectClass(value as new (...arguments_: any[]) => unknown, options, inspect, indent, depth);
        }

        // If it is an object with an anonymous prototype, display it as an object.
        return inspectObject(value as object, options, inspect, indent, depth);
    }

    // last chance to check if it's an object
    if (value === Object(value)) {
        return inspectObject(value as object, options, inspect, indent, depth);
    }

    // We have run out of options! Just stringify the value
    return (options as InternalOptions).stylize(String(value), type);
};

// eslint-disable-next-line @typescript-eslint/no-unsafe-function-type
export const registerConstructor = (constructor: Function, inspector: Inspect): boolean => {
    if (constructorMap.has(constructor)) {
        return false;
    }

    constructorMap.set(constructor, inspector);

    return true;
};

export const registerStringTag = (stringTag: string, inspector: Inspect): boolean => {
    if (stringTag in stringTagMap) {
        return false;
    }

    stringTagMap[stringTag] = inspector;

    return true;
};
