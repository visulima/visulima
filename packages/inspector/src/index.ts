import { internalInspect } from "./internal-inspect";
import type { Options } from "./types";

export { registerConstructor, registerStringTag } from "./internal-inspect";

export const inspect = (value: unknown, options_: Partial<Options> = {}): string => {
    const options = {
        breakLength: Number.POSITIVE_INFINITY,
        compact: 3,
        customInspect: true,
        depth: 5,
        getters: false,
        indent: undefined,
        maxArrayLength: Number.POSITIVE_INFINITY,
        maxStringLength: Number.POSITIVE_INFINITY,
        numericSeparator: true,
        quoteStyle: "single",
        showHidden: false,
        showProxy: false,
        sorted: false,
        stylize: <S extends string>(s: S) => s.toString(),
        ...options_,
    } satisfies Options;

    if ((options.compact === false || options.breakLength !== Number.POSITIVE_INFINITY) && options.indent === undefined) {
        options.indent = 2;
    }

    if (options.indent !== undefined && options.indent !== "\t" && !(Number.parseInt(options.indent as unknown as string, 10) === options.indent && options.indent > 0)) {
        throw new TypeError("option \"indent\" must be \"\\t\", an integer > 0, or `undefined`");
    }

    return internalInspect(value, options, 0, []);
};

export type { Options } from "./types";
