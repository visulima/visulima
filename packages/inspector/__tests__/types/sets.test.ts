import { describe, expect, it } from "vitest";

import { inspect } from "../../src";

describe("sets", () => {
    it("should return `Set {}` for empty sets", () => {
        expect.assertions(1);

        expect(inspect(new Set())).toBe("Set (0) {}");
    });

    it("should return `Set` with indent", () => {
        expect.assertions(6);

        const set = new Set();

        set.add({ a: 1 });
        set.add(["b"]);

        const expectedStringSpaces = ["Set (2) {", "  {", "    a: 1", "  },", "  [ 'b' ]", "}"].join("\n");
        const expectedStringTabs = ["Set (2) {", "	{", "		a: 1", "	},", "	[ 'b' ]", "}"].join("\n");

        expect(inspect(set, { indent: 2 }), 'new Set([{ a: 1 }, ["b"]]) should show size and contents (two)').toBe(expectedStringSpaces);
        expect(inspect(set, { indent: "\t" }), 'new Set([{ a: 1 }, ["b"]]) should show size and contents (tabs)').toBe(expectedStringTabs);

        expect(inspect(new Set(), { indent: 2 }), "empty Set should show as empty (two)").toBe("Set (0) {}");
        expect(inspect(new Set(), { indent: "\t" }), "empty Set should show as empty (tabs)").toBe("Set (0) {}");

        const nestedSet = new Set();

        nestedSet.add(set);
        nestedSet.add(nestedSet);

        const expectedNestedSpaces = ["Set (2) {", "  Set (2) {", "    {", "      a: 1", "    },", "    [ 'b' ]", "  },", "  [Circular]", "}"].join("\n");
        const expectedNestedTabs = ["Set (2) {", "	Set (2) {", "		{", "			a: 1", "		},", "		[ 'b' ]", "	},", "	[Circular]", "}"].join("\n");

        expect(inspect(nestedSet, { indent: 2 }), "Set containing a Set should work (two)").toBe(expectedNestedSpaces);
        expect(inspect(nestedSet, { indent: "\t" }), "Set containing a Set should work (tabs)").toBe(expectedNestedTabs);
    });

    describe("truncate", () => {
        it("returns the full representation when truncate is over string length", () => {
            expect.assertions(1);

            expect(inspect(new Set(["a", "b", "c"]), { truncate: 20 })).toBe("Set (3) { 'a', 'b', 'c' }");
        });

        it("truncates set values longer than truncate (19)", () => {
            expect.assertions(1);

            expect(inspect(new Set(["a", "b", "c"]), { truncate: 19 })).toBe("Set (3) { 'a', …(2) }");
        });

        it("truncates set values longer than truncate (18)", () => {
            expect.assertions(1);

            expect(inspect(new Set(["a", "b", "c"]), { truncate: 18 })).toBe("Set (3) { 'a', …(2) }");
        });

        it("truncates set values longer than truncate (17)", () => {
            expect.assertions(1);

            expect(inspect(new Set(["a", "b", "c"]), { truncate: 17 })).toBe("Set (3) { 'a', …(2) }");
        });

        it("truncates set values longer than truncate (16)", () => {
            expect.assertions(1);

            expect(inspect(new Set(["a", "b", "c"]), { truncate: 16 })).toBe("Set (3) { 'a', …(2) }");
        });

        it("truncates set values longer than truncate (15)", () => {
            expect.assertions(1);

            expect(inspect(new Set(["a", "b", "c"]), { truncate: 15 })).toBe("Set (3) { …(3) }");
        });

        it("truncates set values longer than truncate (14)", () => {
            expect.assertions(1);

            expect(inspect(new Set(["a", "b", "c"]), { truncate: 14 })).toBe("Set (3) { …(3) }");
        });

        it("truncates set values longer than truncate (13)", () => {
            expect.assertions(1);

            expect(inspect(new Set(["a", "b", "c"]), { truncate: 13 })).toBe("Set (3) { …(3) }");
        });

        it("truncates set values longer than truncate (12)", () => {
            expect.assertions(1);

            expect(inspect(new Set(["a", "b", "c"]), { truncate: 12 })).toBe("Set (3) { …(3) }");
        });

        it("truncates set values longer than truncate (11)", () => {
            expect.assertions(1);

            expect(inspect(new Set(["a", "b", "c"]), { truncate: 11 })).toBe("Set (3) { …(3) }");
        });

        it("truncates set values longer than truncate (10)", () => {
            expect.assertions(1);

            expect(inspect(new Set(["a", "b", "c"]), { truncate: 10 })).toBe("Set (3) { …(3) }");
        });

        it("truncates set values longer than truncate (9)", () => {
            expect.assertions(1);

            expect(inspect(new Set(["a", "b", "c"]), { truncate: 9 })).toBe("Set (3) { …(3) }");
        });

        it("truncates set values longer than truncate (8)", () => {
            expect.assertions(1);

            expect(inspect(new Set(["a", "b", "c"]), { truncate: 8 })).toBe("Set (3) { …(3) }");
        });

        it("truncates set values longer than truncate (7)", () => {
            expect.assertions(1);

            expect(inspect(new Set(["a", "b", "c"]), { truncate: 7 })).toBe("Set (3) { …(3) }");
        });

        it("truncates set values longer than truncate (6)", () => {
            expect.assertions(1);

            expect(inspect(new Set(["a", "b", "c"]), { truncate: 6 })).toBe("Set (3) { …(3) }");
        });

        it("truncates set values longer than truncate (5)", () => {
            expect.assertions(1);

            expect(inspect(new Set(["a", "b", "c"]), { truncate: 5 })).toBe("Set (3) { …(3) }");
        });

        it("truncates set values longer than truncate (4)", () => {
            expect.assertions(1);

            expect(inspect(new Set(["a", "b", "c"]), { truncate: 4 })).toBe("Set (3) { …(3) }");
        });

        it("truncates whole array if truncate 3 or less (3)", () => {
            expect.assertions(1);

            expect(inspect(new Set(["a", "b", "c"]), { truncate: 3 })).toBe("Set (3) { …(3) }");
        });

        it("truncates whole array if truncate 3 or less (2)", () => {
            expect.assertions(1);

            expect(inspect(new Set(["a", "b", "c"]), { truncate: 2 })).toBe("Set (3) { …(3) }");
        });

        it("truncates whole array if truncate 3 or less (1)", () => {
            expect.assertions(1);

            expect(inspect(new Set(["a", "b", "c"]), { truncate: 1 })).toBe("Set (3) { …(3) }");
        });
    });
});
