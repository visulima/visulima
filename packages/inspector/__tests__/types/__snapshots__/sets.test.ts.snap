// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sets > should return \`Set\` with indent > nested set indent 2 1`] = `
"Set (2) {
  Set (2) {
    {
      a: 1
    },
    [
      'b'
    ]
  },
  [Circular]
}"
`;

exports[`sets > should return \`Set\` with indent > nested set indent tabs 1`] = `
"Set (2) {
	Set (2) {
		{
			a: 1
		},
		[
			'b'
		]
	},
	[Circular]
}"
`;

exports[`sets > should return \`Set\` with indent > set indent 2 1`] = `
"Set (2) {
  {
    a: 1
  },
  [
    'b'
  ]
}"
`;

exports[`sets > should return \`Set\` with indent > set indent tabs 1`] = `
"Set (2) {
	{
		a: 1
	},
	[
		'b'
	]
}"
`;
